arangod/Cluster/MaintenanceFeature.cpp

```diff
@@ -432,3 +433,159 @@ VPackBuilder  MaintenanceFeature::toVelocyPack() const {
 std::string MaintenanceFeature::toJson(VPackBuilder & builder) {
 } // MaintenanceFeature::toJson
 #endif
+
+std::string const SLASH("/");
+
+arangodb::Result MaintenanceFeature::storeShardError (
+  std::string const& database, std::string const& collection,
+  std::string const& shard, std::shared_ptr<VPackBuffer<uint8_t>> error) {
+
+  std::string key = database + SLASH + collection + SLASH + shard;
+  
+  MUTEX_LOCKER(guard, _seLock);
+  auto const it = _shardErrors.find(key);
+  if (it != _shardErrors.end()) {
+    std::stringstream error;
+    error << "shard " << key << " already has pending error";
+    LOG_TOPIC(DEBUG, Logger::MAINTENANCE) << error.str();
+    return Result(TRI_ERROR_FAILED, error.str());
+  }
+
+  try {
+    _shardErrors.emplace(key,error);
+  } catch (std::exception const& e) {
+    return Result(TRI_ERROR_FAILED, e.what());
+  }
+
+  return Result();
+  
+}
+
+arangodb::Result MaintenanceFeature::shardError (
+  std::string const& database, std::string const& collection,
+  std::string const& shard, std::shared_ptr<VPackBuffer<uint8_t>> error) const {
+
+  std::string key = database + SLASH + collection + SLASH + shard;
+
+  MUTEX_LOCKER(guard, _seLock);
+  auto const it = _shardErrors.find(key);
+  error = (it != _shardErrors.end()) ? it->second : nullptr;
```

Comment:
This has no effect, since the shared_ptr is passed in by value. Change this to a reference.